recipes/openai_textcat.py

```diff
@@ -0,0 +1,535 @@
+        if len(self.examples) >= self.max_examples:
+            self.examples = self.examples[-self.max_examples :]
+
+    def stream_suggestions(
+        self, stream: Iterable[Dict], batch_size: int
+    ) -> Iterable[Dict]:
+        """Get zero-shot or few-shot suggested TextCat annotations from OpenAI.
+
+        Given a stream of input examples, we define a prompt, get a response from OpenAI,
+        and yield each example with their predictions to the output stream.
+        """
+        for batch in _batch_sequence(stream, batch_size):
+            prompts = [
+                self._get_textcat_prompt(
+                    eg["text"], label=self.label, examples=self.examples
+                )
+                for eg in batch
+            ]
+            responses = self._get_textcat_response(prompts, delay=0)
+            for eg, prompt, response in zip(batch, prompts, responses):
+                if self.verbose:
+                    rich.print(Panel(prompt, title="Prompt to OpenAI"))
+                eg["openai"] = {"prompt": prompt, "response": response}
+                if self.verbose:
+                    rich.print(Panel(response, title="Response from OpenAI"))
+                yield eg
+
+    def format_suggestions(
+        self, stream: Iterable[Dict], *, nlp: Language
+    ) -> Iterable[Dict]:
+        """Parse the examples in the stream and set up labels
+        to display in the Prodigy UI.
+        """
+        stream = prodigy.components.preprocess.add_tokens(nlp, stream, skip=True)  # type: ignore
+        for example in stream:
+            example = copy.deepcopy(example)
+            response = self._parse_response(example["openai"]["response"])
+            example["answer"] = response["answer"] == "accept"
+            example["meta"]["reason"] = response["reason"]
+            yield example
+
+    def filter_suggestions(
+        self, stream: Iterable[Dict], negative_bias: float
+    ) -> Iterable[Dict]:
+        """Filter the examples based on some negative_bias"""
+        for example in stream:
+            if example.get("answer") != "accept":
+                rng = random.uniform(0, 1)
+                if rng <= negative_bias:
+                    eg = example
+                else:
+                    continue
+            else:
+                eg = example
+            yield prodigy.util.set_hashes(eg)
+
+    def _get_textcat_prompt(
+        self, text: str, label: str, examples: List[PromptExample]
+    ) -> str:
+        """Generate a prompt for binary text categorization."""
+        return self.prompt_template.render(text=text, label=label)
+
+    def _get_textcat_response(self, prompts: List[str], delay: int = 0) -> List[str]:
+        headers = {
+            "Authorization": f"Bearer {self.openai_api_key}",
+            "OpenAI-Organization": self.openai_api_org,
+            "Content-Type": "application/json",
+        }
+        time.sleep(delay)
+        r = _retry429(
```

Comment:
While running the recipe I also got this error: 

```
future: <Task finished name='Task-32' coro=<RequestResponseCycle.run_asgi() done, defined at /home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py:394> exception=HTTPStatusError("Server error '503 Service Unavailable' for url 'https://api.openai.com/v1/completions'\nFor more information check: https://httpstatuses.com/503")>
Traceback (most recent call last):
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 399, in run_asgi
    self.logger.error(msg, exc_info=exc)
  File "/usr/lib/python3.8/logging/__init__.py", line 1463, in error
    self._log(ERROR, msg, args, **kwargs)
  File "/usr/lib/python3.8/logging/__init__.py", line 1577, in _log
    self.handle(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 1586, in handle
    if (not self.disabled) and self.filter(record):
  File "/usr/lib/python3.8/logging/__init__.py", line 807, in filter
    result = f.filter(record)
  File "cython_src/prodigy/util.pyx", line 145, in prodigy.util.ServerErrorFilter.filter
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/uvicorn/protocols/http/h11_impl.py", line 396, in run_asgi
    result = await app(self.scope, self.receive, self.send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/uvicorn/middleware/proxy_headers.py", line 45, in __call__
    return await self.app(scope, receive, send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/fastapi/applications.py", line 208, in __call__
    await super().__call__(scope, receive, send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/applications.py", line 112, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/middleware/errors.py", line 181, in __call__
    raise exc from None
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/middleware/errors.py", line 159, in __call__
    await self.app(scope, receive, _send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/middleware/cors.py", line 86, in __call__
    await self.simple_response(scope, receive, send, request_headers=headers)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/middleware/cors.py", line 142, in simple_response
    await self.app(scope, receive, send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/middleware/base.py", line 25, in __call__
    response = await self.dispatch_func(request, self.call_next)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/prodigy/app.py", line 205, in reset_db_middleware
    response = await call_next(request)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/middleware/base.py", line 45, in call_next
    task.result()
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/middleware/base.py", line 38, in coro
    await self.app(scope, receive, send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/exceptions.py", line 82, in __call__
    raise exc from None
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/exceptions.py", line 71, in __call__
    await self.app(scope, receive, sender)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/routing.py", line 580, in __call__
    await route.handle(scope, receive, send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/routing.py", line 241, in handle
    await self.app(scope, receive, send)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/routing.py", line 52, in app
    response = await func(request)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/fastapi/routing.py", line 226, in app
    raw_response = await run_endpoint_function(
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/fastapi/routing.py", line 161, in run_endpoint_function
    return await run_in_threadpool(dependant.call, **values)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/starlette/concurrency.py", line 40, in run_in_threadpool
    return await loop.run_in_executor(None, func, *args)
  File "/usr/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/prodigy/app.py", line 434, in get_session_questions
    return _shared_get_questions(req.session_id, excludes=req.excludes)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/prodigy/app.py", line 399, in _shared_get_questions
    tasks = controller.get_questions(session_id=session_id, excludes=excludes)
  File "cython_src/prodigy/core.pyx", line 221, in prodigy.core.Controller.get_questions
  File "cython_src/prodigy/core.pyx", line 222, in prodigy.core.Controller.get_questions
  File "cython_src/prodigy/components/feeds.pyx", line 379, in prodigy.components.feeds.Feed.get_batch
  File "cython_src/prodigy/components/feeds.pyx", line 323, in prodigy.components.feeds.Feed._enqueue_tasks
  File "cython_src/prodigy/components/stream.pyx", line 170, in prodigy.components.stream.Stream.__next__
  File "cython_src/prodigy/components/stream.pyx", line 174, in prodigy.components.stream.Stream.__next__
  File "cython_src/prodigy/components/filters.pyx", line 80, in filter_inputs
  File "recipes/openai_textcat.py", line 211, in filter_suggestions
    for example in stream:
  File "recipes/openai_textcat.py", line 198, in format_suggestions
    for example in stream:
  File "cython_src/prodigy/components/preprocess.pyx", line 164, in add_tokens
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/spacy/language.py", line 1539, in pipe
    for doc in docs:
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/spacy/language.py", line 1583, in pipe
    for doc in docs:
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/spacy/language.py", line 1580, in <genexpr>
    docs = (self._ensure_doc(text) for text in texts)
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/spacy/language.py", line 1529, in <genexpr>
    docs_with_contexts = (
  File "cython_src/prodigy/components/preprocess.pyx", line 157, in genexpr
  File "recipes/openai_textcat.py", line 182, in stream_suggestions
    responses = self._get_textcat_response(prompts, delay=0)
  File "recipes/openai_textcat.py", line 249, in _get_textcat_response
    r.raise_for_status()
  File "/home/vincent/Development/prodigy-demos/venv/lib/python3.8/site-packages/httpx/_models.py", line 736, in raise_for_status
    raise HTTPStatusError(message, request=request, response=self)
httpx.HTTPStatusError: Server error '503 Service Unavailable' for url 'https://api.openai.com/v1/completions'
For more information check: https://httpstatuses.com/503
```

We might want to retry these as well?

Reply:
I also got this error every time I tried batch sizes of 25.

```
httpx.HTTPStatusError: Client error '400 Bad Request' for url 'https://api.openai.com/v1/completions'
```

Might be a good idea to print the response from OpenAI here. The json response usually displays the error that it's dealing with.

Reply:
Still working on this, but just noting that I saw this!

Reply:
Updates: https://github.com/explosion/prodigy-openai-recipes/pull/50#issuecomment-1396563787